File: sqldelight-runtime/src/main/java/com/squareup/sqldelight/EnumColumnAdapter.java
/*
 * Copyright (C) 2016 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.sqldelight;

import android.support.annotation.NonNull;

/** A {@link ColumnAdapter} which maps the enum class {@code T} to a string in the database. */
public final class EnumColumnAdapter<T extends Enum<T>> implements ColumnAdapter<T, String> {
  public static <T extends Enum<T>> EnumColumnAdapter<T> create(Class<T> cls) {
    if (cls == null) throw new NullPointerException("cls == null");
    return new EnumColumnAdapter<>(cls);
  }

  private final Class<T> cls;

  private EnumColumnAdapter(Class<T> cls) {
    this.cls = cls;
  }

  @Override @NonNull public T decode(String databaseValue) {
    return Enum.valueOf(cls, databaseValue);
  }

  @Override public String encode(@NonNull T value) {
    return value.name();
  }
}

File: sqldelight-runtime/src/commonMain/kotlin/com/squareup/sqldelight/EnumColumnAdapter.kt
/*
 * Copyright (C) 2016 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.sqldelight

/** A [ColumnAdapter] which maps the enum class `T` to a string in the database. */
class EnumColumnAdapter<T : Enum<T>> @PublishedApi internal constructor(
  private val enumValues: Array<out T>
) : ColumnAdapter<T, String> {
  override fun decode(databaseValue: String) = enumValues.first { it.name == databaseValue }

  override fun encode(value: T) = value.name
}

@Suppress("FunctionName") // Emulating a constructor.
inline fun <reified T : Enum<T>> EnumColumnAdapter(): EnumColumnAdapter<T> {
  return EnumColumnAdapter(enumValues())
}
